### PR TITLE
network: Only insert global addresses into the DHT.

### PR DESCRIPTION
**Pull request by twittner**
Currently every address reported via libp2p-identify is inserted into the DHT which thus contains a multitude of unreachable addresses such as from 127.0.0.0/8 or 10.0.0.0/8.

Issue #5099 suggested a dedicated service over UDP to gauge the reachability of an address, which would however incur extra I/O costs and be of limited use.

As an alternative and simpler tactic, this PR only allows global IP addresses to be inserted into the DHT unless an explicit command-line flag `--allow-non-global-addresses-in-dht` is given or a node is started with `--dev`. This opt-in behaviour is meant to allow site-local networks to still make use of a DHT.

### CODE REVIEW

**Comment by tomaka** [comment]
You probably have to enable the flag for tests to pass. The tests use `/memory/...` addresses.

File: client/network/src/discovery.rs [hunk 1]
```diff
@@ -123,6 +126,12 @@ impl DiscoveryConfig {
 		self
 	}
 
+	/// Should non-global addresses be inserted to the DHT?
+	pub fn allow_non_globals_in_dht(&mut self, value: bool) -> &mut Self {
+		self.allow_non_globals_in_dht = value;
+		self
+	}
```
Comment (mxinden) [inline client/network/src/discovery.rs:133]: Given that it defaults to false, why not:

```suggestion
	pub fn allow_non_globals_in_dht(&mut self) -> &mut Self {
		self.allow_non_globals_in_dht = true;
		self
	}
```

File: client/network/src/discovery.rs [hunk 2]
```diff
@@ -298,6 +314,23 @@ impl DiscoveryBehaviour {
...
+	/// Can the given `Multiaddr` be put into the DHT?
+	///
+	/// This test is successful only for global IP addresses and DNS names.
+	//
+	// NB: Currently all DNS names are allowed and no check for TLD suffixes is done
+	// because the set of valid domains is highly dynamic and would require frequent
+	// updates, for example by utilising publicsuffix.org or IANA.
+	pub fn can_add_to_dht(&self, addr: &Multiaddr) -> bool {
+		let ip = match addr.iter().next() {
+			Some(Protocol::Ip4(ip)) => IpNetwork::from(ip),
```
Comment (mxinden) [inline client/network/src/discovery.rs:327]: I guess once the `std` API stabilizes we can use https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.is_global.

Comment (twittner) [reply client/network/src/discovery.rs:327]: Yes, I pulled in `ip_network` precisely for this reason.

File: client/network/src/discovery.rs (hunk 1 above)
Comment (twittner) [reply client/network/src/discovery.rs:133]: The current implementation is consistent with `allow_private_ipv4`, so we should then change both. I do not think it makes much of a difference so unless you (or anyone else) feels strongly about it I would probably leave it as is.

**Comment by mxinden** [comment]
:+1: for `B1-clientnoteworthy`. Maybe someone is using Substrate in a private IPv4 network.

**Comment by arkpar** [comment]
CLI option name could be a bit more user friendly. Something like `--discover-local` - Enable peer discovery on local networks. "DHT" is an implementation detail.

**Comment by twittner** [comment]
> CLI option name could be a bit more user friendly. Something like `--discover-local` - Enable peer discovery on local networks. "DHT" is an implementation detail.

I would be fine changing it to this. The meaning of "discover-local" overlaps perhaps slightly with mDNS but since we have a separate option to disable mDNS I guess it is fine.

**Comment by tomaka** [comment]
Please merge master for the CI to pass.

**Pull request merged**